File: master.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>
#include <time.h>
#include <stdbool.h>

#define OUT_DEGREE 10
#define CHUNK_SIZE 1000000

typedef enum {NORMAL, CPU, GPU} ActivityType;

typedef struct {
    int64_t target;
    float weight;
    float data;       // stores user-provided data
    bool control;
    ActivityType activity;
} Connection;

typedef struct {
    float state;
    float polarity;
    Connection outs[OUT_DEGREE];
} Irobit;

float rand_float() {
    return ((float)rand()/RAND_MAX)*2.0f - 1.0f;
}

void build_chunk(Irobit *chunk, int chunk_size, int64_t total_irobits) {
    for(int i=0; i<chunk_size; i++) {
        chunk[i].state = 0.0f;
        chunk[i].polarity = (rand()%2)?1.0f:-1.0f;
        for(int j=0;j<OUT_DEGREE;j++){
            chunk[i].outs[j].target = rand()%total_irobits;
            chunk[i].outs[j].weight = rand_float();
            chunk[i].outs[j].data = rand_float();  // store user data
            chunk[i].outs[j].control = rand()%2;
            int r = rand()%100;
            if(r<85) chunk[i].outs[j].activity=NORMAL;
            else if(r<95) chunk[i].outs[j].activity=CPU;
            else chunk[i].outs[j].activity=GPU;
        }
    }
}

float tanh_f(float x){
    return (expf(x)-expf(-x))/(expf(x)+expf(-x));
}

void step_chunk(Irobit *chunk, int chunk_size){
    float *inputs = (float*)calloc(chunk_size,sizeof(float));
    for(int i=0;i<chunk_size;i++){
        float s = chunk[i].state;
        for(int j=0;j<OUT_DEGREE;j++){
            Connection c = chunk[i].outs[j];
            float factor = c.data; // user data affects propagation
            if(c.activity==CPU) inputs[i] += s*c.weight*1.2f*factor;
            else if(c.activity==GPU) inputs[i] += s*c.weight*1.5f*factor;
            else inputs[i] += s*c.weight*factor;
        }
    }
    for(int i=0;i<chunk_size;i++){
        chunk[i].state = tanh_f(inputs[i]*chunk[i].polarity);
    }
    free(inputs);
}

int main(){
    srand(time(NULL));
    int64_t total_irobits = 1000000000000LL;
    int chunk_idx = 0;

    while(1){
        Irobit *chunk = malloc(sizeof(Irobit)*CHUNK_SIZE);
        build_chunk(chunk, CHUNK_SIZE, total_irobits);

        for(int i=0;i<50 && i<CHUNK_SIZE;i++) chunk[i].state = 1.0f;

        for(int step=0; step<10; step++){
            step_chunk(chunk, CHUNK_SIZE);
        }

        free(chunk);
        chunk_idx++;
        printf("Processed chunk %d\n", chunk_idx);
    }
    return 0;
}

File: master.py
import numpy as np
import random
import time

# ===== Configuration =====
N_IROBITS = 10**12   # 1 trillion iRobits
OUT_DEGREE = 10
CHUNK_SIZE = 10**6       # simulate in chunks
POLARITY_RATIO = 0.5
STEPS = 10               # steps per chunk update
ACTIVITY_TYPES = ['NORMAL', 'CPU', 'GPU']
SEED = 42

random.seed(SEED)
np.random.seed(SEED)

# Connection dtype
dtype_connection = np.dtype([
    ('target', np.int64),
    ('weight', np.float32),
    ('data', np.float32),    # now stores user data for all
    ('control', np.bool_),
    ('activity', 'U4')       # 'CPU', 'GPU', 'NORMAL'
])

# ===== Build Chunk =====
def build_chunk(chunk_size):
    n = chunk_size
    connections = np.zeros((n, OUT_DEGREE), dtype=dtype_connection)
    polarity = np.ones(n, dtype=np.float32)

    neg_count = int((1.0-POLARITY_RATIO)*n)
    if neg_count>0:
        neg_indices = np.random.choice(n, neg_count, replace=False)
        polarity[neg_indices] = -1.0

    for i in range(n):
        for j in range(OUT_DEGREE):
            target = random.randint(0, N_IROBITS-1)
            weight = random.uniform(-1.0, 1.0)
            # User-provided data stored in all connections
            data = random.uniform(-1.0, 1.0)  # can replace with user input
            control = random.choice([True, False])
            activity = random.choices(ACTIVITY_TYPES, weights=[0.85,0.1,0.05])[0]
            connections[i,j] = (target, weight, data, control, activity)
    return connections, polarity, np.zeros(n, dtype=np.float32)

# ===== Step Simulation =====
def step_chunk(states, connections, polarity):
    n = states.size
    inputs = np.zeros(n, dtype=np.float32)
    for i in range(n):
        s = states[i]
        for j in range(connections.shape[1]):
            c = connections[i,j]
            # Use stored data in computation
            factor = c.data  # factor based on stored user data
            if c.activity == 'CPU':
                inputs[i] += s * c.weight * 1.2 * factor
            elif c.activity == 'GPU':
                inputs[i] += s * c.weight * 1.5 * factor
            else:
                inputs[i] += s * c.weight * factor
    new_states = np.tanh(inputs * polarity)
    return new_states

# ===== Main Loop =====
def main():
    chunk_idx = 0
    while True:
        print(f"Initializing chunk {chunk_idx}")
        conns, pol, states = build_chunk(CHUNK_SIZE)

        # Inject initial values
        inject_indices = np.random.choice(CHUNK_SIZE, min(50, CHUNK_SIZE), replace=False)
        states[inject_indices] = 1.0

        for step in range(STEPS):
            states = step_chunk(states, conns, pol)
            if step % 2 == 0:
                top_idx = np.argsort(-np.abs(states))[:5]
                print(f"Chunk {chunk_idx} Step {step} top: {[(i, states[i]) for i in top_idx]}")
        
        chunk_idx += 1

if __name__ == "__main__":
    main()
